### ISSUE TITLE
Problems using native editors via $env:EDITOR with Git for Windows.

### ISSUE THREAD

**Issue by rkitover** [git-for-windows/git#3629]
### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```console
git version 2.34.1.windows.1
cpu: x86_64
built from commit: 2ca94ab318509b3c271e82889938816bad76dfea
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```console
Microsoft Windows [Version 10.0.22526.1000]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```console
Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I am using the new Windows Terminal with PowerShell Core.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```powershell
$env:EDITOR = 'C:/tools/neovim/Neovim/bin/nvim.exe'
git commit -a --verbose --signoff -S
```

 - What did you expect to occur after running these commands?

I expected Neovim to start with my commit message in the same mode as it starts when run from the PowerShell prompt.

 - What actually happened instead?

Neovim started with terminal support for Cygwin, with much worse colors etc..

Running the following inside Neovim shows:

```vim
:echo $TERM
" cygwin
:echo &term
" builtin_cygwin
```

, while starting it from PowerShell shows:

```vim
:echo $TERM
" vtpcon
:echo &term
" builtin_vtpcon
```

The issue is that `$env:TERM` is set to `cygwin` when Neovim launches from Git, causing it to use its builtin Cygwin terminal support which is much worse than the native `vtpcon` support for the new Windows console.

I opened an issue related to this in the neovim repository:

https://github.com/neovim/neovim/issues/17021

There are two workarounds that I found for this issue, the first is creating a `.bat` file and setting it as `$env:EDITOR`.

This my `~/.local/bin/nvim.bat`:

```dosbatch
@echo off
set TERM=
/tools/neovim/neovim/bin/nvim %*
```

. Now, the following would work correctly:

```powershell
$env:EDITOR = 'C:/Users/rkitover/.local/bin/nvim.bat'
git commit -a --verbose --signoff -S
```

. The other workaround is adding the following to `~/.gitconfig`:

```gitconfig
[core]
    editor = sh -c "'"TERM= c:/tools/neovim/Neovim/bin/nvim.exe \"$@\""'" --
```

.

### Other Problems with `$env:EDITOR` on Windows

- A normal Windows path with backslashes in `$env:EDITOR` does not work at all:

```powershell
$env:EDITOR = 'C:\tools\neovim\Neovim\bin\nvim.exe'
git commit -a --verbose --signoff -S
```
```console
hint: Waiting for your editor to close the file... C:\tools\neovim\Neovim\bin\nvim.exe: C:toolsneovimNeovimbinnvim.exe: command not found
error: There was a problem with the editor 'C:\tools\neovim\Neovim\bin\nvim.exe'.
Please supply the message using either -m or -F option.
```

**Comment by dscho** [comment]
> The issue is that `$env:TERM` is set to `cygwin` when Neovim launches from Git, causing it to use its builtin Cygwin terminal support which is much worse than the native `vtpcon` support for the new Windows console.

Isn't the real problem, then, that Neovim's defaults for the `cygwin` terminal are suboptimal?

**Comment by rkitover** [comment]
> Isn't the real problem, then, that Neovim's defaults for the `cygwin` terminal are suboptimal?

This too, the new cygwin terminal support has almost full support for the new Windows console. However, they may want to support older systems. Further complicating this is that this is the native version of Neovim and not the Cygwin or MSYS2 version (MSYS2 is just Cygwin with a small patch for parameter-destroying.)

I should mention that the regular version of native vim does not have this problem, it ignores `$env:TERM` and uses its Windows console support.

I will bring this up in the Neovim issue I linked.

Regardless, I consider Git modifying `$env:TERM` a bug, if mine is empty I want it to stay empty when my editor or any external program launches.

Native Windows apps generally don't use `$env:TERM`, and there is no standard for this at present.

**Comment by dscho** [comment]
If native Windows programs don't use `TERM`, why do we even bother with this discussion?

**Comment by rkitover** [comment]
I tried setting:

```powershell
$env:TERM = 'vtpcon'` 
```

and that works correctly with Neovim, and Git does not modify it in that case.

**Comment by dscho** [comment]
You could use one of the two approaches outlined in https://stackoverflow.com/questions/59733731/how-to-detect-if-running-in-the-new-windows-terminal or https://github.com/microsoft/terminal/issues/6269 to detect in your PowerShell profile that it is running inside a Windows Terminal, and then set `$env:TERM = 'vtpcon':

1. use the presence of `$env:WT_SESSION` (although this might be a false positive if you spawn, say, a MinTTY window from your session)
2. walk the `ParentProcessId` chain to figure out whether the current session is a (transitive) child process of the Windows Terminal (it _does_ run the same risk of a false positive, of course).

Of course, you could also change your Windows Terminal PowerShell Profile to execute with a startup command that sets this (I _think_ `-NoExit -Command '$env:TERM="vtpcon"'` might work, but I haven't verified this).

**Comment by rkitover** [comment]
I took a different approach, I am setting:

```powershell
$env:TERM = 'xterm-256color'
$env:COLORTERM = 'truecolor'
```

, in my `$profile`. This is currently the most widely used method on Linux as well. The `$env:COLORTERM` is not necessary in my case, but I set it just in case.

Both the Windows Terminal and versions of conhost made in the last many years support this just fine. All features I care about including 24 bit color work fine in Neovim. Regular vim also works fine, to the extent that it does on Windows, nano works fine.

So I have a solution for my problem and for the document I am maintaining.

But that does not solve the issue here, which is `$env:TERM` being set to `cygwin` when it is unset for native, non-Cygwin (including MSYS2 and related) apps.

I consider this a bug, if you don't, that's for you to decide.

Thank you for your work on this project.

**Comment by rkitover** [comment]
One of the terminal developers confirmed that these are good settings to use, and this works fine with the vim you ship with Git, other than the mouse not working, but that is likely a terminal issue. The mouse does not work in the native regular vim either.

**Comment by rkitover** [comment]
I did some experiments and it seems that your version of Cygwin is setting `$TERM` correctly, the problem seems to be with how Git is launching `$env:EDITOR`.

![git bash $TERM](https://i.imgur.com/WYdDHez.png)

I could take a look at this if you like, this is in `editor.c`.

**Comment by dscho** [comment]
> I could take a look at this if you like, this is in `editor.c`.

If you mean to say that https://github.com/git-for-windows/git/blob/v2.35.0.windows.1/editor.c sets `TERM=cygwin`, I would like to learn how it does that.

**Comment by rkitover** [comment]
> I also noticed that, and despite looking through the wiki and searching for "windows" in git-config(1), I could not find where this is documented.. @dscho is it documented somewhere ?

My guess is that `EDITOR` is invoked with bash without proper quoting which treats backslashes as escapes. It seems that MSYS2/Cygwin bash is capable if launching Windows paths if they are properly quoted, for example, this will work:

```bash
'C:\Windows\notepad.exe'
```
.

**Comment by rkitover** [comment]
> I agree that it should set TERM=xterm-256color according to what I understand of [that code](https://github.com/git-for-windows/msys2-runtime/blob/b600e8ead500aef55e23810c2e630d9be46f3a4c/winsup/cygwin/wincap.cc#L455-L471). I'm on 1909, so 10.0.18363.2037, and so I should get the wincap defined for 1903. However, I'm also seeing "TERM=cygwin" on my side, though with native Vim (not Neovim). How would I go about getting GDB set up to step inside wincap.cc ? I did look through [Debugging Git](https://github.com/git-for-windows/git/wiki/Debugging-Git) but it does not mention how to debug the runtime...

Did you see my example screenshot? The runtime works correctly, the problem is in Git.

I was going to look in `editor.c` before I got side-tracked by looking at a bunch of cmake stuff, I'd suggest you look there.

**Comment by rkitover** [comment]
Here are the commands you can test in case someone deleted my screenshot:

```powershell
ri env:TERM,env:COLORTERM -ea ignore

&'/program files/Git/bin/bash' -c 'echo $TERM'

&'/program files/Git/bin/bash' -c 'c:/tools/neovim/neovim/bin/nvim -c ''call writefile($TERM), fnamemodify("~",":p") . "git-bash-nvim-term.txt")|q''' && gc ~/git-bash-nvim-term.txt
```

**Comment by rkitover** [comment]
I mistyped that, sorry, the second command is:

```powershell
&'/program files/Git/bin/bash' -c 'c:/tools/neovim/neovim/bin/nvim -c ''call writefile([$TERM], fnamemodify("~",":p") . "git-bash-nvim-term.txt")|q'''; gc ~/git-bash-nvim-term.txt
```

, I will write an equivalent for regular vim in a few minutes.

**Comment by rkitover** [comment]
On the subject of Windows paths in `$env:EDITOR`, I just realized that this works:

```powershell
$env:EDITOR = "'C:/tools/neovim/Neovim/bin/nvim.exe'"
```

, note the extra single quotes.

We just need a good solution to do this automatically somehow.

**Comment by phil-blain** [comment]
The second command still does not work for me (sorry, I don't no much about Vim nor Powershell!). If it matters, this is under Windows Powershell (i.e. 5.1). The following command "works":
~~~pwsh
 &"$(where.exe git)\..\..\bin\bash.exe" -c "'$(where.exe vim)' -c 'call writefile([$TERM], `"`"term.txt`"`")|q'"; cat term.txt
~~~
but the resulting file `term.txt` is empty...
(I have both Git and Vim installed in my account so I generalized your example to not hardcode installed locations).

**Comment by phil-blain** [comment]
OK, I was missing a backtick before TERM:
~~~pwsh
 &"$(where.exe git)\..\..\bin\bash.exe" -c "'$(where.exe vim)' -c 'call writefile([`$TERM], `"`"term.txt`"`")|q'"; cat term.txt
~~~
This does show `xterm-256color` for me as well.

But, is the environment setup the same when running `/path/to/Git/bin/bash.exe` directly vs running via  `/path/to/Git/cmd/git.exe`, I don't know...

**Comment by rkitover** [comment]
Never mind, I did this wrong.

I forgot to run:

```powershell
ri env:TERM,env:COLORTERM -ea ignore
```

, having done this, the `$TERM` is set to `cygwin` in both cases.

So you are right about the problem being with Git in general.

**Comment by rkitover** [comment]
It also appears to be a problem with the `git.exe` executable specifically, trying with the included MSYS2 `awk`:

```powershell
&'/Program Files/Git/usr/bin/awk' 'BEGIN { print(ENVIRON["TERM"]); exit }'
```

, I also get `xterm-256color`.

**Comment by rkitover** [comment]
So I found the problem:

https://github.com/git-for-windows/git/blob/main/compat/mingw.c#L3365

**Comment by phil-blain** [comment]
Indeed! It dates back to https://github.com/git-for-windows/git/commit/baea068d677fae92d7903e984cf93bbd5195a000 and this logic was not touched since them as far as I could see.

**Comment by phil-blain** [comment]
> Of course, you could also change your Windows Terminal PowerShell Profile to execute with a startup command that sets this (I _think_ `-NoExit -Command '$env:TERM="vtpcon"'` might work, but I haven't verified this).

A correct syntax for this seems to be:
~~~pwsh
# From Powershell itself
powershell.exe -NoLogo -NoExit -Command '$env:TERM=\"xterm-256color\"'
# or 
powershell.exe -NoLogo -NoExit -Command '$env:TERM=''xterm-256color'''
# or
powershell.exe -NoLogo -NoExit -Command "`$env:TERM='xterm-256color'"
~~~

and in Terminal's `settings.json`:
~~~json
"commandline": "powershell.exe -NoLogo -NoExit -Command \"$env:TERM='xterm-256color'\"",
~~~

**Comment by rkitover** [comment]
As I already mentioned, the most sane way to work around this right now is to put the following in your PowerShell `$profile`:

```powershell
$env:TERM = 'xterm-256color'
$env:COLORTERM = 'truecolor'
```

this will work with any Windows terminal or console made in the last few years.

See: https://github.com/microsoft/terminal/issues/12174

@phil-blain Thank you for helping me debug these, you want to take up making the two PRs for this or should I?

**Comment by phil-blain** [comment]
@rkitover you can go ahead :) feel free to tag / cc me though.

**Comment by dscho** [comment]
> So I found the problem:
> 
> https://github.com/git-for-windows/git/blob/main/compat/mingw.c#L3365

My most pressing question is: can we easily detect that we're running inside a Windows Terminal and use `xterm-256color` in that case instead of `cygwin`?

Previously (and I think, elsewhere), I mentioned that `WT_SESSION` can be stale too easily, e.g. when you open a MinTTY from within a Windows Terminal session.

However, the longer I think about it the more I get convinced that this is still the best we could do... @phil-blain @rkitover any takers?

**Comment by dscho** [comment]
> Reading the message for [baea068](https://github.com/git-for-windows/git/commit/baea068d677fae92d7903e984cf93bbd5195a000) which was merged in 2014, it seems that Cygwin did not do this at the time and that's why this change was introduced.
> 
> It seems it would be reasonable to drop this change

Sounds reasonable. Would you mind testing this?

**Comment by rkitover** [comment]
No problem.

**Comment by rimrul** [comment]
> Reading the message for https://github.com/git-for-windows/git/commit/baea068d677fae92d7903e984cf93bbd5195a000 which was merged in 2014, it seems that Cygwin did not do this at the time

This is probably true, but I think it just chose `cygwin` if nothing else was set back then.

> and that's why this change was introduced.

I don't think it's the reason. The commit talks about running `git` from `cmd.exe` and `color.c`, so I think this is about native `git.exe` builtin commands assuming `TERM=dumb` without this patch.

**Comment by rkitover** [comment]
@rimrul 

Perhaps we could enclose it in:

```c
#ifndef __CYGWIN__
...
#endif
```
?

**Comment by rimrul** [comment]
I don't think we compile `compat/mingw.c` in situations where `__CYGWIN__` would be defined.

In fact the earliest versions of this used `TERM=winansi`, not `TERM=cygwin`

https://github.com/msysgit/git/commit/29a3963484d3ee943a0b3cc7f5e4eb355b562826

**Comment by rkitover** [comment]
@rimrul @dscho 

Since this is not running under cygwin, should we just implement the same check that cygwin does for this [here](https://github.com/git-for-windows/git/issues/3629#issuecomment-1024061897)?

**Comment by rimrul** [comment]
> @rimrul @dscho 
> 
> Since this is not running under cygwin, should we just implement the same check that cygwin does for this [here](https://github.com/git-for-windows/git/issues/3629#issuecomment-1024061897)?

That could work.

**Comment by dscho** [comment]
> Well I tried:
> 
> ```c
> #ifndef __CYGWIN__
> ```
> 
> , and it did not fix the problem.

Right, that cannot fix the problem because we build `git.exe` as MINGW executable. However, when we shell out, we use the MSYS2 Bash, which is an MSYS executable, and that's where the MSYS2 runtime comes in and all those ramifications regarding the `TERM` variable come from.

> > Since this is not running under cygwin, should we just implement the same check that cygwin does for this [here](https://github.com/git-for-windows/git/issues/3629#issuecomment-1024061897)?
> 
> That could work.

That link points to the MSYS2 runtime, where we have the `wincap.has_con_24bit_colors()` function. In `git.exe`, we have no access to that, and would therefore have to implement it ourselves.

The `wincap` stuff is defined [here](https://github.com/git-for-windows/msys2-runtime/blob/HEAD/winsup/cygwin/wincap.cc), and as you can see, it defines the capabilities as constants depending on the Windows version. It would appear that [Windows 10 1703 is the first to support 24-bit colors](https://github.com/git-for-windows/msys2-runtime/blob/00534e8184011ab37a3f63c501d881d488ebe542/winsup/cygwin/wincap.cc#L261). And this version [corresponds to the build number 15063](https://github.com/git-for-windows/msys2-runtime/blob/00534e8184011ab37a3f63c501d881d488ebe542/winsup/cygwin/wincap.cc#L465-L466).

So what we could do is [imitate existing code looking for the build number](https://github.com/git-for-windows/git/blob/v2.35.1.windows.2/compat/mingw.c#L3715) like so:

```c
	if (!getenv("TERM"))
		setenv("TERM", GetVersion() >> 16 >= 15063 ? "xterm-256color" : "cygwin", 1);
```

**Comment by rimrul** [comment]
That's correct (when calling git from cmd/powershell).

**Comment by hehuahuang** [comment]
Matthias Aßhauer ***@***.***> 於 2022年2月15日 週二 23:35 寫道：

**Comment by dscho** [comment]
Does this ticket need anything apart from https://github.com/git-for-windows/git/pull/3751 (which I just merged, please verify using the next [snapshot](https://wingit.blob.core.windows.net/files/index.html) once it is done [building](https://dev.azure.com/git-for-windows/git/_build/results?buildId=98211&view=results))? Assuming that it does not, I will tentatively close it.

**Comment by rkitover** [comment]
@dscho I have just verified that the snapshot fixes the issue with the `$env:TERM`, I will open a separate issue regarding `$env:EDITOR` quoting.
